// puppeteer-core/lib/cjs/puppeteer/common/Errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TargetCloseError = exports.UnsupportedOperation = exports.ProtocolError = exports.TouchError = exports.TimeoutError = exports.PuppeteerError = void 0;
class PuppeteerError extends Error {
    constructor(message, options) {
        super(message, options);
        this.name = this.constructor.name;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
}
exports.PuppeteerError = PuppeteerError;
class TimeoutError extends PuppeteerError {
}
exports.TimeoutError = TimeoutError;
class TouchError extends PuppeteerError {
}
exports.TouchError = TouchError;
class ProtocolError extends PuppeteerError {
    #code;
    #originalMessage = '';
    set code(code) {
        this.#code = code;
    }
    get code() {
        return this.#code;
    }
    set originalMessage(originalMessage) {
        this.#originalMessage = originalMessage;
    }
    get originalMessage() {
        return this.#originalMessage;
    }
}
exports.ProtocolError = ProtocolError;
class UnsupportedOperation extends PuppeteerError {
}
exports.UnsupportedOperation = UnsupportedOperation;
class TargetCloseError extends ProtocolError {
}
exports.TargetCloseError = TargetCloseError;
