// puppeteer-core/lib/cjs/puppeteer/common/Puppeteer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Puppeteer = void 0;
const BrowserConnector_js_1 = require("./BrowserConnector.js");
const CustomQueryHandler_js_1 = require("./CustomQueryHandler.js");
class Puppeteer {
    static customQueryHandlers = CustomQueryHandler_js_1.customQueryHandlers;
    static registerCustomQueryHandler(name, queryHandler) {
        return this.customQueryHandlers.register(name, queryHandler);
    }
    static unregisterCustomQueryHandler(name) {
        return this.customQueryHandlers.unregister(name);
    }
    static customQueryHandlerNames() {
        return this.customQueryHandlers.names();
    }
    static clearCustomQueryHandlers() {
        return this.customQueryHandlers.clear();
    }
    _isPuppeteerCore;
    _changedBrowsers = false;
    constructor(settings) {
        this._isPuppeteerCore = settings.isPuppeteerCore;
        this.connect = this.connect.bind(this);
    }
    connect(options) {
        return (0, BrowserConnector_js_1._connectToBrowser)(options);
    }
}
exports.Puppeteer = Puppeteer;
