// puppeteer-extra-plugin-stealth/evasions/chrome.csi/index.js
'use strict'

const { PuppeteerExtraPlugin } = require('puppeteer-extra-plugin')

const withUtils = require('../_utils/withUtils')

class Plugin extends PuppeteerExtraPlugin {
  constructor(opts = {}) {
    super(opts)
  }

  get name() {
    return 'stealth/evasions/chrome.csi'
  }

  async onPageCreated(page) {
    await withUtils(page).evaluateOnNewDocument(utils => {
      if (!window.chrome) {
        
        
        Object.defineProperty(window, 'chrome', {
          writable: true,
          enumerable: true,
          configurable: false, 
          value: {} 
        })
      }

      
      if ('csi' in window.chrome) {
        return 
      }

      
      if (!window.performance || !window.performance.timing) {
        return
      }

      const { timing } = window.performance

      window.chrome.csi = function() {
        return {
          onloadT: timing.domContentLoadedEventEnd,
          startE: timing.navigationStart,
          pageT: Date.now() - timing.navigationStart,
          tran: 15 
        }
      }
      utils.patchToString(window.chrome.csi)
    })
  }
}

module.exports = function(pluginConfig) {
  return new Plugin(pluginConfig)
}
