// chromium-bidi/lib/esm/bidiMapper/modules/permissions/PermissionsProcessor.js
import { InvalidArgumentException, } from '../../../protocol/protocol.js';
export class PermissionsProcessor {
    #browserCdpClient;
    constructor(browserCdpClient) {
        this.#browserCdpClient = browserCdpClient;
    }
    async setPermissions(params) {
        try {
            const userContextId = params['goog:userContext'] ||
                params.userContext;
            await this.#browserCdpClient.sendCommand('Browser.setPermission', {
                origin: params.origin,
                browserContextId: userContextId && userContextId !== 'default'
                    ? userContextId
                    : undefined,
                permission: {
                    name: params.descriptor.name,
                },
                setting: params.state,
            });
        }
        catch (err) {
            if (err.message ===
                `Permission can't be granted to opaque origins.`) {
                
                
                return {};
            }
            throw new InvalidArgumentException(err.message);
        }
        return {};
    }
}
