// chromium-bidi/lib/cjs/protocol-parser/generated/webdriver-bidi-permissions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Permissions = exports.PermissionsCommandSchema = void 0;


const zod_1 = __importDefault(require("zod"));
exports.PermissionsCommandSchema = zod_1.default.lazy(() => Permissions.SetPermissionSchema);
var Permissions;
(function (Permissions) {
    Permissions.PermissionDescriptorSchema = zod_1.default.lazy(() => zod_1.default.object({
        name: zod_1.default.string(),
    }));
})(Permissions || (exports.Permissions = Permissions = {}));
(function (Permissions) {
    Permissions.PermissionStateSchema = zod_1.default.lazy(() => zod_1.default.enum(['granted', 'denied', 'prompt']));
})(Permissions || (exports.Permissions = Permissions = {}));
(function (Permissions) {
    Permissions.SetPermissionSchema = zod_1.default.lazy(() => zod_1.default.object({
        method: zod_1.default.literal('permissions.setPermission'),
        params: Permissions.SetPermissionParametersSchema,
    }));
})(Permissions || (exports.Permissions = Permissions = {}));
(function (Permissions) {
    Permissions.SetPermissionParametersSchema = zod_1.default.lazy(() => zod_1.default.object({
        descriptor: Permissions.PermissionDescriptorSchema,
        state: Permissions.PermissionStateSchema,
        origin: zod_1.default.string(),
        userContext: zod_1.default.string().optional(),
    }));
})(Permissions || (exports.Permissions = Permissions = {}));
