// puppeteer-extra-plugin-user-preferences/index.js
'use strict'

const merge = require('deepmerge')

const { PuppeteerExtraPlugin } = require('puppeteer-extra-plugin')

class Plugin extends PuppeteerExtraPlugin {
  constructor(opts = {}) {
    super(opts)
    this._userPrefsFromPlugins = {}

    const defaults = {
      userPrefs: {}
    }

    this._opts = Object.assign(defaults, opts)
  }

  get name() {
    return 'user-preferences'
  }

  get requirements() {
    return new Set(['runLast', 'dataFromPlugins'])
  }

  get dependencies() {
    return new Set(['user-data-dir'])
  }

  get data() {
    return [
      {
        name: 'userDataDirFile',
        value: {
          target: 'Profile',
          file: 'Preferences',
          contents: JSON.stringify(this.combinedPrefs, null, 2)
        }
      }
    ]
  }

  get combinedPrefs() {
    return merge(this._opts.userPrefs, this._userPrefsFromPlugins)
  }

  async beforeLaunch(options) {
    this._userPrefsFromPlugins = merge.all(
      this.getDataFromPlugins('userPreferences').map(d => d.value)
    )
    this.debug('_userPrefsFromPlugins', this._userPrefsFromPlugins)
  }
}

module.exports = function(pluginConfig) {
  return new Plugin(pluginConfig)
}
