// puppeteer-core/lib/cjs/puppeteer/api/Input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Touchscreen = exports.Mouse = exports.MouseButton = exports.Keyboard = void 0;
const Errors_js_1 = require("../common/Errors.js");
const incremental_id_generator_js_1 = require("../util/incremental-id-generator.js");
class Keyboard {
    constructor() { }
}
exports.Keyboard = Keyboard;
exports.MouseButton = Object.freeze({
    Left: 'left',
    Right: 'right',
    Middle: 'middle',
    Back: 'back',
    Forward: 'forward',
});
class Mouse {
    constructor() { }
}
exports.Mouse = Mouse;
class Touchscreen {
    idGenerator = (0, incremental_id_generator_js_1.createIncrementalIdGenerator)();
    touches = [];
    constructor() { }
    removeHandle(handle) {
        const index = this.touches.indexOf(handle);
        if (index === -1) {
            return;
        }
        this.touches.splice(index, 1);
    }
    async tap(x, y) {
        const touch = await this.touchStart(x, y);
        await touch.end();
    }
    async touchMove(x, y) {
        const touch = this.touches[0];
        if (!touch) {
            throw new Errors_js_1.TouchError('Must start a new Touch first');
        }
        return await touch.move(x, y);
    }
    async touchEnd() {
        const touch = this.touches.shift();
        if (!touch) {
            throw new Errors_js_1.TouchError('Must start a new Touch first');
        }
        await touch.end();
    }
}
exports.Touchscreen = Touchscreen;
